libwnbd/wnbd_ioctl.cpp

```diff
@@ -139,6 +144,367 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
+        if (ERROR_NO_MORE_ITEMS != Status) {
+            LogError(
+                "Failed to locate device with hardware id: %s. Error: %d. Error message: %s",
+                WNBD_HARDWAREID, Status, win32_strerror(Status).c_str());
+        }
+        return Status;
+    }
+
+    return ERROR_SUCCESS;
+}
+
+DWORD RemoveAllDevices(BOOLEAN Force)
+{
+    DWORD BufferSize = 0;
+    DWORD Status = 0;
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+
+    WNBD_OPTION_VALUE OptValue = { WnbdOptBool };
+    /* Disallow new mappings so we can remove all current mappings */
+    OptValue.Data.AsBool = FALSE;
+    Status = WnbdSetDrvOpt("NewMappingsAllowed", &OptValue, FALSE);
+    if (Status) {
+        goto exit;
+    }
+
+    Status = WnbdList(ConnList, &BufferSize);
+    if (!BufferSize) {
+        goto exit;
+    }
+    ConnList = (PWNBD_CONNECTION_LIST)calloc(1, BufferSize);
+    if (NULL == ConnList) {
+        Status = ERROR_NOT_ENOUGH_MEMORY;
+        LogError(
+            "Failed to allocate %d bytes.", BufferSize);
+        goto exit;
+    }
+    Status = WnbdList(ConnList, &BufferSize);
+    if (Status) {
+        goto exit;
+    }
+    CHAR* InstanceName;
+    for (ULONG index = 0; index < ConnList->Count; index++) {
+        InstanceName = ConnList->Connections[index].Properties.InstanceName;
+        WNBD_REMOVE_OPTIONS RemoveOptions = { 0 };
+        RemoveOptions.Flags.HardRemoveFallback = TRUE;
+        RemoveOptions.SoftRemoveTimeoutMs = 30 * 1000;
+        RemoveOptions.SoftRemoveRetryIntervalMs = 1 * 1000;
+        RemoveOptions.Flags.HardRemove = Force;
+        Status = WnbdRemoveEx(InstanceName, &RemoveOptions);
+        if (Status) {
+            goto exit;
+        }
+    }
+
+exit:
+    if (ConnList) {
+        free(ConnList);
+    }
+    return Status;
+}
+
+DWORD WnbdUninstallDriver(PBOOL RebootRequired, BOOLEAN Force)
+{
+    DWORD Status = ERROR_SUCCESS;
+    HDEVINFO AdapterDevHandle = INVALID_HANDLE_VALUE;
+    SP_DEVINFO_DATA DevInfoData;
+    DevInfoData.cbSize = sizeof DevInfoData;
+
+    PWNBD_CONNECTION_LIST ConnList = NULL;
+    Status = RemoveAllDevices(Force);
+    if (Status) {
+        goto exit;
+    }
+
+    while (Status == ERROR_SUCCESS) {
+        Status = FindWnbdAdapterDevice(&AdapterDevHandle, &DevInfoData);
+        if (ERROR_SUCCESS != Status) {
+            goto exit;
+        }
+
+        Status = RemoveWnbdDevice(AdapterDevHandle, &DevInfoData, RebootRequired);
+
+        if (INVALID_HANDLE_VALUE != AdapterDevHandle) {
+            if (!SetupDiDestroyDeviceInfoList(AdapterDevHandle)) {
+                Status = GetLastError();
+                LogError("SetupDiDestroyDeviceInfoList failed with "
+                    "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+            }
+        }
+
+        if (ERROR_SUCCESS != Status) {
+            LogError(
+                "Error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
```

Comment:
An error without any context isn't very useful. Plus, it's redundant since all the above errors are logged so we can just remove it.